File: tests/testthat/test-dataviz.R
# The dataviz_friendly flag and the measurements behind it.

test_that("the catalogue carries the dataviz columns", {
  p <- palettes()
  expect_type(p$dataviz_friendly, "logical")
  expect_false(anyNA(p$dataviz_friendly))
  for (col in c("min_delta_e", "min_delta_e_cvd", "min_delta_e_white")) {
    expect_type(p[[col]], "double")
    expect_false(anyNA(p[[col]]), info = col)
    expect_true(all(p[[col]] >= 0), info = col)
  }
})

test_that("some palettes pass and some fail", {
  # a flag that is constant would be useless
  p <- palettes()
  expect_gt(sum(p$dataviz_friendly), 0)
  expect_gt(sum(!p$dataviz_friendly), 0)
})

test_that("the flag agrees with the measurements and the shipped thresholds", {
  skip_if_not_installed("jsonlite")
  path <- system.file("extdata", "japanesecolors-palettes.json", package = "japanesecolors")
  skip_if(path == "", "canonical JSON not installed")
  thresholds <- jsonlite::fromJSON(path, simplifyVector = TRUE)$dataviz$thresholds

  p <- palettes()
  expected <- p$min_delta_e >= thresholds$min_delta_e &
    p$min_delta_e_cvd >= thresholds$min_delta_e_cvd &
    p$min_delta_e_white >= thresholds$min_delta_e_white

  expect_identical(p$dataviz_friendly, expected)
})

test_that("CVD separation does not substantially exceed normal-vision separation", {
  # Sanity check on the simulation: dichromacy should not pull colours apart.
  # The Machado matrices are fitted linear approximations rather than strict
  # projections onto a dichromat gamut, so with sRGB clamping a distance can
  # creep up marginally (retro_tri_11 goes 37.0 to 37.1). A large increase would
  # mean the simulation is wrong, so allow slack but not much.
  p <- palettes()
  expect_true(all(p$min_delta_e_cvd <= p$min_delta_e + 2))
})

test_that("dataviz_friendly filters the catalogue", {
  yes <- palettes(dataviz_friendly = TRUE)
  no <- palettes(dataviz_friendly = FALSE)
  expect_true(all(yes$dataviz_friendly))
  expect_false(any(no$dataviz_friendly))
  expect_identical(nrow(yes) + nrow(no), nrow(palettes()))

  expect_identical(palette_names(dataviz_friendly = TRUE), yes$id)
  expect_error(palettes(dataviz_friendly = "yes"), "TRUE or FALSE")
  expect_error(palettes(dataviz_friendly = NA), "TRUE or FALSE")
})

test_that("dataviz_friendly combines with the other filters", {
  sub <- palettes("retro", n = 4, dataviz_friendly = TRUE)
  expect_true(all(sub$collection == "retro"))
  expect_true(all(sub$n_colors == 4L))
  expect_true(all(sub$dataviz_friendly))
})

test_that("every flagged palette really is separable", {
  # guards against the flag drifting away from the data it summarises
  thresholds_cvd <- 10
  for (i in seq_len(nrow(palettes(dataviz_friendly = TRUE)))) {
    row <- palettes(dataviz_friendly = TRUE)[i, ]
    expect_gte(row$min_delta_e_cvd, thresholds_cvd, label = row$id)
  }
})

File: python/tests/test_dataviz.py
"""The dataviz_friendly flag and the measurements behind it."""

from __future__ import annotations

import pytest

import japanesecolors as jc

METRICS = ("min_delta_e", "min_delta_e_cvd", "min_delta_e_white")


def test_catalogue_carries_the_dataviz_fields():
    for row in jc.palettes():
        assert isinstance(row["dataviz_friendly"], bool)
        for key in METRICS:
            assert isinstance(row[key], (int, float)), key
            assert row[key] >= 0


def test_some_pass_and_some_fail():
    rows = jc.palettes()
    assert 0 < sum(r["dataviz_friendly"] for r in rows) < len(rows)


def test_flag_agrees_with_measurements_and_thresholds():
    t = jc.DATAVIZ["thresholds"]
    for row in jc.palettes():
        expected = (
            row["min_delta_e"] >= t["min_delta_e"]
            and row["min_delta_e_cvd"] >= t["min_delta_e_cvd"]
            and row["min_delta_e_white"] >= t["min_delta_e_white"]
        )
        assert row["dataviz_friendly"] is expected, row["id"]


def test_cvd_separation_does_not_substantially_exceed_normal_vision():
    """Sanity check on the simulation: dichromacy should not pull colours apart.

    The Machado matrices are fitted linear approximations rather than strict
    projections onto a dichromat gamut, so with sRGB clamping a distance can
    creep up marginally -- retro_tri_11 goes 37.0 to 37.1. A large increase
    would mean the simulation is wrong, so the check allows slack but not much.
    """
    for row in jc.palettes():
        assert row["min_delta_e_cvd"] <= row["min_delta_e"] + 2.0, row["id"]


def test_filtering():
    yes = jc.palettes(dataviz_friendly=True)
    no = jc.palettes(dataviz_friendly=False)
    assert all(r["dataviz_friendly"] for r in yes)
    assert not any(r["dataviz_friendly"] for r in no)
    assert len(yes) + len(no) == len(jc.palettes())
    assert jc.palette_names(dataviz_friendly=True) == [r["id"] for r in yes]

    with pytest.raises(TypeError, match="True, False or None"):
        jc.palettes(dataviz_friendly="yes")


def test_filter_combines_with_others():
    rows = jc.palettes("retro", n=4, dataviz_friendly=True)
    assert all(
        r["collection"] == "retro" and r["n_colors"] == 4 and r["dataviz_friendly"]
        for r in rows
    )


def test_r_and_python_agree_on_the_flag():
    """The scores are computed once and shared, so the packages must match."""
    import json
    import pathlib

    shipped = pathlib.Path(jc.__file__).resolve().parents[3] / "inst" / "extdata" / (
        "japanesecolors-palettes.json"
    )
    if not shipped.exists():
        pytest.skip("not running from the source tree")
    with shipped.open(encoding="utf-8") as fh:
        r_rows = {p["id"]: p for p in json.load(fh)["palettes"]}

    for row in jc.palettes():
        r = r_rows[row["id"]]
        assert r["dataviz_friendly"] == row["dataviz_friendly"], row["id"]
        for key in METRICS:
            assert r[key] == pytest.approx(row[key]), f"{row['id']}/{key}"


def test_known_good_reference_palette_would_pass():
    """Sanity-check the thresholds against Okabe-Ito, which they are tuned to."""
    t = jc.DATAVIZ["thresholds"]
    assert t["min_delta_e"] <= 21.7
    assert t["min_delta_e_cvd"] <= 11.1
